### PR TITLE
Throw when attempting to store empty blobs

### LINKED ISSUE #156: Fails to store empty blob
With a schema like this:
```
schema.addTableDefinition("blobs", new TableDefinition() { {
                rowName();
                rowComponent("id", ValueType.STRING);

                columns();
                column("blob", "b", ValueType.BLOB);
            }
        });
```
Storage and retrieval works fine in general, but if you store an empty byte array (i.e. length zero), then retrieval will fail to find anything. It will give an Optional.empty when you look up that row.

I tried to force Atlas to be aware of empty blobs as such:
```
schema.addTableDefinition("blobs", new TableDefinition() { {
                rowName();
                rowComponent("id", ValueType.STRING);

                columns();
                column("blob", "b", ValueType.BLOB);
                column("timestamp", "t", ValueType.VAR_LONG);
            }
        });
```
Then the row would be found (not Optional.absent), but the blob itself would be null.

### PR DESCRIPTION
**Pull request by gbrova**
Fixes #156

Currently cell deletion in Atlas works by inserting an empty value; for this reason, a deleted cell and an empty cell are indistinguishable. For now, we do not support empty values. 

This PR improves usability by failing loudly -- inserting an empty row will now throw an `IllegalArgumentException`, instead of just deleting the cell or logging a warning (which users in the field may not notice).

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/524)

### CODE REVIEW

**Comment by gbrova** [comment]
Open question: I'm not sure whether to include the [third commit](https://github.com/palantir/atlasdb/pull/524/commits/00ca677ae1c7f79f066d6ae00bd538ac217218f3) and allow empty entries in non-empty rows. My initial thought in adding it was that a more lenient policy might minimize breaking changes, and it specifically allows the workaround described in #156. I think it's safe because we can know a row exists if it has at least one non-empty cell. However, I'm happy to remove it if it's not worth the added complexity.

File: docs/source/release_notes/release-notes.rst [hunk 1]
```diff
@@ -23,6 +23,25 @@ Changelog
...
+=======
+
+.. list-table::
+    :widths: 5 40
+    :header-rows: 1
+
+    *   - Type
+        - Change
+
+    *   - |changed|
```
Comment (joelea) [inline docs/source/release_notes/release-notes.rst]: This should be breaking and we should create that category if it doesn't already exist.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 2]
```diff
@@ -655,6 +655,13 @@ public void testWriteWriteConflictsDeletedThrow() {
         }
     }
 
+    @Test (expected = IllegalArgumentException.class)
+    public void disallowPutOnEmptyObject() {
+        final Cell cell = Cell.create("row1".getBytes(), "column1".getBytes());
+        Transaction t1 = txManager.createNewTransaction();
+        t1.put(TABLE, ImmutableMap.of(cell, new byte[0]));
```
Comment (joelea) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: Can we use EMPTY_BYTE_ARRAY like the deleted lines above? I like it having a name.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    private RowResult<byte[]> getRowResult(final TableReference table, final byte[] rowName) {
+        return txManager.runTaskWithRetry(new TransactionTask<RowResult<byte[]>, RuntimeException>() {
+                @Override
+                public RowResult<byte[]> execute(Transaction t) throws RuntimeException {
+                    return t.getRows(table, ImmutableSet.of(rowName), ColumnSelection.all()).get(rowName);
+                }
+            });
+    }
+
+    private static final String BLOB_ONLY_TABLE_NAME = "blobsOnly";
```
Comment (joelea) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionIntegrationTest.java]: The constants should be at the top of the class.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    private static final String BLOB_AND_VALUE_TABLE_NAME = "blobsAndLongs";
+
+    private static final Cell blobCell = Cell.create("id".getBytes(Charsets.UTF_8), "blob".getBytes(Charsets.UTF_8));
```
Comment (joelea) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionIntegrationTest.java]: constant so should be THIS_CASE

File: docs/source/release_notes/release-notes.rst (hunk 1 above)
Comment (gbrova) [inline docs/source/release_notes/release-notes.rst]: Added a "Breaking" category for now. 

However, I'm not clear on what the difference between "Changed" and "Breaking" is; how do you feel about renaming "Changed" -> "Breaking" instead?

**Comment by gbrova** [comment]
Answering my own question -- after discussing with @joelea we decided to keep things as simple as possible. So, `Transaction.put(...)` will throw an `IllegalArgumentException` when _any_ of the cells is empty, even if they are part of a non-empty row.

File: docs/source/release_notes/release-notes.rst (hunk 1 above)
Comment (joelea) [inline docs/source/release_notes/release-notes.rst]: If we never find a use for changed then I'm happy with that - but a big red breaking is necessary in my view.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 5]
```diff
@@ -943,10 +943,24 @@ protected boolean shouldDeleteAndRollback() {
...
 
+    private void ensureNoEmptyValues(Map<Cell, byte[]> values) {
+        for (Entry<Cell, byte[]> cellEntry : values.entrySet()) {
+            if (cellEntry.getValue().length == 0) {
+                throw new IllegalArgumentException("AtlasDB does not currently support inserting empty (zero-byte) values.");
+            }
+        }
+    }
+
     public void put(TableReference tableRef, Map<Cell, byte[]> values, long ttlDuration, TimeUnit ttlUnit) {
```
Comment (joelea) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: Given this is public, aren't we only disallowing adding empty values on one method? It seems weird that the 2 different puts have different contracts.

Comment (gbrova) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: Good point. This method isn't part of the API and as far as I can tell should not be public - changed to private.

**Comment by joelea** [comment]
Looks good, lets wait for the KVS release to put this in - avoid the breaking changes.

**Comment by clockfort** [comment]
flagging as something we need to check usages of when we pull this back over the wall @tmgordeeva

**Pull request merged**